### PR TITLE
BF: gh-3997, never use the worktree as reference when diffing against a commit

### PR DESCRIPTION
**Pull request by mih**
Story in gh-3997

### CODE REVIEW

File: datalad/core/local/tests/test_diff.py [hunk 1]
```diff
@@ -481,3 +481,31 @@ def test_diff_nonexistent_ref_unicode(path):
...
+    # now perform a diff for the last commit, there is one file that remained
+    # identifical
+    ds.repo.call_git(['checkout', 'test'])
+    res = ds.diff(fr='master~1', to='master')
+    # under no circumstances can there be any reports on deleted files
+    # because we never deleted anything
+    assert_result_count(res, 0, state='deleted')
+    # the identical file must be reported clean
+    assert_result_count(
+        ds.diff(fr='master~1', to='master'),
```
Comment (kyleam) [inline datalad/core/local/tests/test_diff.py]: Nit: You could reuse `res` from above.

**Comment by mih** [comment]
This must go into 0.12, hence quick feedback is appreciated.

File: datalad/support/gitrepo.py [hunk 2]
```diff
@@ -3546,8 +3546,8 @@ def _get_cache_key(label, paths, ref, untracked=None):
                 if to_state_r['type'] != 'dataset':
                     # no change in git record, and no change on disk
                     props = dict(
-                        state='clean' if f.exists() or \
-                              f.is_symlink() else 'deleted',
+                        state='clean' if to is not None or (f.exists() or \
+                              f.is_symlink()) else 'deleted',
```
Comment (kyleam) [inline datalad/support/gitrepo.py]: Okay, so at this point we know that the reported object ids for this file are identical in the `to` and `from` records.  If `to` is None, we're comparing to the working tree and a deleted file will still have an identical id, so we need to check whether the file is gone before declaring it clean.  This working tree check is irrelevant and wrong if `to` is a ref.

Makes sense.

File: datalad/core/local/tests/test_diff.py (hunk 1 above)
Comment (mih) [reply datalad/core/local/tests/test_diff.py]: A true... spedd accuracy trade-off ;-)

File: datalad/support/gitrepo.py (hunk 2 above)
Comment (mih) [reply datalad/support/gitrepo.py]: Thx, that is indeed the reasoning. I will take the liberty and place it into a comment ;-)

Comment (kyleam) [reply datalad/support/gitrepo.py]: Ah, no need!   I was just working through it.

Comment (kyleam) [reply datalad/support/gitrepo.py]: But I see you already did it, which is also ok :)

Comment (mih) [reply datalad/support/gitrepo.py]: Done nevertheless!

File: datalad/core/local/tests/test_diff.py (hunk 1 above)
Comment (mih) [reply datalad/core/local/tests/test_diff.py]: Done!

**Pull request merged**